wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java

```diff
@@ -110,13 +133,109 @@ protected boolean doHandleRequest(final Request request) {
 		boolean changed = value != null || current != null;
 
 		if (changed) {
-			setData(value);
-		}
+			// Reset validation fields
+			resetValidationState();
+			
+			// if fileType is supplied then validate it
+			boolean validFileType;
+			if (isSetFileTypes()) {
+				validFileType = FileValidationUtil.validateFileType(value, getFileTypes());
+				addFileTypeValidation(validFileType, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_TYPE, 
+						getFileTypes().toArray(new Object[getFileTypes().size()]));
+			} else {
+				validFileType = true;
+			}
+			
+			// if fileSize is supplied then validate it
+			boolean validFileSize;
+			if (isSetFileSize()) {
+				validFileSize = FileValidationUtil.validateFileSize(value, getMaxFileSize());
+				addFileSizeValidation(validFileSize, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_SIZE, 
+						Util.readableFileSize(value.getSize()), Util.readableFileSize(getMaxFileSize()));
+			} else {
+				validFileSize = true;
+			}
+			
+			// if file is valid, the update data
+			if (validFileSize && validFileType) {
+				setData(value);
+			} else {
```

Comment:
If there was a file previously loaded, this will need to be cleared (ie set null) and return "true" as changed.